Userland/Utilities/cat.cpp

```diff
@@ -37,13 +41,43 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
 
     TRY(Core::System::pledge("stdio"));
 
+    size_t line_count = 1;
     Array<u8, 32768> buffer;
     for (auto const& file : files) {
         while (!file->is_eof()) {
             auto const buffer_span = TRY(file->read_some(buffer));
-            out("{:s}", buffer_span);
+            if (show_lines) {
+                line_count = output_buffer_with_line_numbers(line_count, buffer_span);
+            } else {
+                out("{:s}", buffer_span);
+            }
         }
     }
 
     return files.size() != paths.size();
 }
+
+size_t output_buffer_with_line_numbers(size_t line_count, ReadonlyBytes buffer_span)
+{
+    auto buffer_itter = buffer_span.begin();
+    while (!buffer_itter.is_end()) {
+        // to know if the next itter is the end
+        auto curr_value = *buffer_itter;
+        buffer_itter++;
+
+        if (line_count == 1 && (buffer_itter == buffer_span.begin())) {
+            out("   {: >6}\t{:c}", line_count, curr_value);
+        } else if (curr_value != '\n') {
+            out("{:c}", curr_value);
+        } else {
+            out("{:c}", curr_value);
+            if (!buffer_itter.is_end()) {
```

Comment:
One issue that I can see here is that you're testing for the end of a relatively small buffer, not the end of the file. If the 32k buffer ends by `\n`, you will be missing a line count.

To fix that, I would take another approach:
Let's declare a `struct` that store the state, with the line count and a `bool` `display_line_number`.
The struct is passed to the function and also returned from it (as you did with `line_count`).
Then `output_buffer_with_line_numbers` can simply be:
```cpp
for (auto const c : buffer) {
	if (display_line_number) {
		out("{: >6}\t", line_count);
		display_line_number = false;
	}
	if (c == '\n')
		display_line_number = true;
	out("{:c}", c);
}
```

Reply:
Yea, sorry about the bad code I was rushing as I haven't had much time this week to work on projects. Will have another go at it on Sunday. Thanks for the help though.